File: app/src/main/java/dev/mina/parcelableillusion/SecondActivity.java
package dev.mina.parcelableillusion;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;

import static dev.mina.parcelableillusion.FirstActivity.FIRST_CHILD_EXTRA;
import static dev.mina.parcelableillusion.FirstActivity.SECOND_CHILD_EXTRA;


class SecondActivity extends AppCompatActivity {

    private ChildOne firstChild;
    private ChildTwo secondChild;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_second);
        Button retrieveChildOne = findViewById(R.id.retrieve_child_one);
        retrieveChildOne.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                firstChild = getIntent().getParcelableExtra(FIRST_CHILD_EXTRA);
            }
        });
        Button retrieveChildTwo = findViewById(R.id.retrieve_child_two);
        retrieveChildTwo.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                secondChild = getIntent().getParcelableExtra(SECOND_CHILD_EXTRA);
            }
        });
    }
}

File: app/src/main/java/dev/mina/parcelableillusion/SecondActivity.kt
package dev.mina.parcelableillusion

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import kotlinx.android.synthetic.main.second_layout.*

class SecondActivity : AppCompatActivity() {


    lateinit var firstChild: ChildOne
    lateinit var secondChild: ChildTwo

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_second)
        retrieve_child_one.setOnClickListener {
            firstChild = intent.getParcelableExtra<ChildOne>(FIRST_CHILD_EXTRA)
        }
        retrieve_child_two.setOnClickListener {
            secondChild = intent.getParcelableExtra<ChildTwo>(SECOND_CHILD_EXTRA)
        }
    }
}
